Add Army Value statistics/graph

ubitux opened the pull request:
This feature makes possible to follow the army value over time in a game, useful for analysis in the competitive scene (typically to spot when good/bad trades happen).

There are a bunch of units for which I'm unsure whether I should count them as part of the army or not. Following is the list of my doubts, open to discussion. Note: some of the unit in RA are redundant in other mods (such as medic).

RA:
    engie? -> no
    medic? -> no
    mechanic? -> no
    spy? -> yes
    minelayer? -> yes
    mobile gap gen? -> yes
    mobile radar jammer? -> yes
    general? -> yes
    thief? -> yes
    hijacker? -> yes
    transport? (ship) -> no
    chinook? -> no

CNC:
    dino? -> no
    landing craft? -> no

D2K:
    orni? -> no
    thumper? -> no
    saboteur? -> yes

TS:
    mutant hijacker? -> yes

Here is a screenshot on how it looks in observer mode. There is also a new column in the Combat tab.

![2018-10-31-223605_1920x1080_scrot](https://user-images.githubusercontent.com/34467/47823395-68d61d00-dd68-11e8-8072-21eb567528a7.png)

ubitux committed:
Add Army value to player statistics

To be accounted as army, the unit needs to have
UpdatesPlayerStatistics.AddToArmyValue to true (false by default)

ubitux committed:
RA: set UpdatesPlayerStatistics.AddToArmyValue where appropriate

ubitux committed:
CNC: set UpdatesPlayerStatistics.AddToArmyValue where appropriate

ubitux committed:
D2K: set UpdatesPlayerStatistics.AddToArmyValue where appropriate

ubitux committed:
TS: set UpdatesPlayerStatistics.AddToArmyValue where appropriate

ubitux committed:
Observer: display army value in stats widget

ubitux committed:
Observer: display army value in a new dedicated graph tab

pchote commented on OpenRA.Mods.Common/Traits/Player/PlayerStatistics.cs:
```diff
@@ -106,9 +120,13 @@ public void ResolveOrder(Actor self, Order order)
+	public class UpdatesPlayerStatisticsInfo : TraitInfo<UpdatesPlayerStatistics>
+	{
+		[Desc("Count as army value in statistics")]
+		public bool CountAsArmy = false;
```
Maybe clearer to call this `AddToArmyValue`?

pchote commented on OpenRA.Mods.Common/Traits/Player/PlayerStatistics.cs:
```diff
@@ -133,6 +151,46 @@ void INotifyKilled.Killed(Actor self, AttackInfo e)
 				attackerStats.KillsCost += cost;
 				defenderStats.DeathsCost += cost;
+
+				var countForArmy = self.Info.TraitInfo<UpdatesPlayerStatisticsInfo>().CountAsArmy;
```
While it would have a negligble effect on performance, it is our style convention to cache the traits `Info` object in an `info` field.  Please change `public class UpdatesPlayerStatisticsInfo : TraitInfo<UpdatesPlayerStatistics>` above to `public class UpdatesPlayerStatisticsInfo : ITraitInfo<UpdatesPlayerStatistics>` and override the `Create` method to call a new `public UpdatesPlayerStatistics(UpdatesPlayerStatisticsInfo info)` ctor that assigns info to a `readonly UpdatesPlayerStatisticsInfo info;` for use here and in Crated.

pchote reviewed (commented):
Code changes look reasonable, just a couple of minor nits.

Can we please get some player testing here?

FRenzy-OpenRA commented:
What is the reasoning behind removing these specific units from the feature ?

I'd suggest enabling them all by default. For instance, medics and mechanics often constitute a significant bulk of Allied armies. Removing them will flaw the comparison in AvS matchup.

ubitux commented:
> What is the reasoning behind removing these specific units from the feature ?
> 
> I'd suggest enabling them all by default. For instance, medics and mechanics often constitute a significant bulk of Allied armies. Removing them will flaw the comparison in AvS matchup.

It followed the basic rule of thumb "any unit that impacts directly the enemy (by reducing enemy health, speed, vision, ...)". As a result, support units were out of this. I don't mind adding them, but then should transport units be in as well? What about the engie? I have no definite opinion on the matter, but a consensus on where to draw the line needs to be made. What I don't want though, is to have the harvesters counted in.

FRenzy-OpenRA commented:
On engies, harvesters, naval transports : not a problem to not count them. 

Chinooks : I'm more cautious on this one, as their Soviet transport counterpart (APC) is weaponed and will be counted. By precaution, I'd say count it in ? 
But in that case, and for consistency, naval transports would be counted as well.

ubitux commented:
> On engies, harvesters, naval transports : not a problem to not count them.
> 
> Chinooks : I'm more cautious on this one, as their Soviet transport counterpart (APC) is weaponed and will be counted. By precaution, I'd say count it in ?

I don't know, again it's all about where you draw the line: look at engie for instance, you could argue that they do affect the enemy player if used as a mean to capture their building. Chinooks never do any damage, but they do if they crash, so should they be counted in? Even harvesters could be counted in since they can be used to tank and crush infantry. 

So yeah again, the current patchset is where *I* drew the line, but obviously don't mind adjusting it to a certain extent (see remark on the harvs in my previous post) in one direction or another. That said, since the issue is debatable in many aspects, I'd prefer if a definite consensus was decided before I make any change.

pchote commented on OpenRA.Mods.Common/Traits/Player/PlayerStatistics.cs:
```diff
@@ -106,9 +120,13 @@ public void ResolveOrder(Actor self, Order order)
+	}
 
-	public class UpdatesPlayerStatistics : INotifyKilled
+	public class UpdatesPlayerStatistics : INotifyKilled, INotifyCreated, INotifyOwnerChanged
```
I expect you will also need to implement `INotifyActorDisposing` so that you can clean up when actors are removed without being explicitly killed.

pchote replied:
You may need to track a flag inside this trait to keep track of whether it is currently counted towards the value, to avoid double (un)counting.

ubitux replied:
What am actually I supposed to do in `INotifyActorDisposing`? How can I trigger it?

ubitux replied:
I've been trying to do that but failed. I'm getting "Exception of type `System.InvalidOperationException`: Attempted to get trait from destroyed object (player 4 (not in world))". I'm not familiar with the language and engine so I'm probably missing something but it seems that I'm getting the notification too late when the context is not in a valid state to poke into the trait anymore.

pchote commented:
I expect that it is going to take too long to form a consensus here, so we're going to have to bump this to Next + 1.

pchote commented:
There is still time to get this in for the playtest if a consensus can be found and reviews done in time.  Choosing not to block the playtest over it doesn't have to mean pushing it out of the playtest completely.

abcdefg30 commented on OpenRA.Mods.Common/Traits/Player/PlayerStatistics.cs:
```diff
@@ -133,6 +160,36 @@ void INotifyKilled.Killed(Actor self, AttackInfo e)
+			if (self.Info.HasTraitInfo<ValuedInfo>() && info.AddToArmyValue)
+			{
+				var playerStats = self.Owner.PlayerActor.Trait<PlayerStatistics>();
+				var cost = self.Info.TraitInfo<ValuedInfo>().Cost;
```
Why don't we cache this instead of querying it each time? I suppose trying to query this in `INotifyActorDisposing` is causing your issue, so that would go away as well with caching.

pchote replied:
Yeah, that would indeed do it.

ubitux replied:
Caching the `cost` works but I'm not sure caching the `playerStats` will work as expected. I'm not familiar with how it works in the language but it feels like it's going to be an orphan reference or something...

pchote replied:
There is no problem with caching `playerStats`. You just need to make sure to update the reference on owner change.

ubitux commented:
OK so, I updated the branch with the following changes:

- `PlayerStatistics.cs`: cache of `UpdatesPlayerStatisticsInfo` and `cost`
- `PlayerStatistics.cs`: honor `INotifyActorDisposing.Disposing()`
- `PlayerStatistics.cs`: key is now named `UpdatesPlayerStatistics.AddToArmyValue` (`true`/`false`)
- mods: simplify a lot the Traits changes by applying AddToArmyValue=true to all soldiers (and cyborg for TS), vehicles, planes and ships, with the 3 exceptions for vehicles: MCV, harvs, and supply truck. Doing so makes the diffs much smallers.
- rebased on `bleed`

Note: I haven't retested all the mods again, so please double checks if more units need exceptions.

ubitux commented:
rebased & ping

pchote reviewed (commented):
A few minor code comments:

ubitux commented on OpenRA.Mods.Common/Traits/Player/PlayerStatistics.cs:
```diff
@@ -128,12 +161,41 @@ void INotifyKilled.Killed(Actor self, AttackInfo e)
+
+		void INotifyOwnerChanged.OnOwnerChanged(Actor self, Player oldOwner, Player newOwner)
+		{
+			if (self.Owner.WinState != WinState.Undefined)
```
Are you referring to that particular opening condition? I just copied it from `INotifyKilled.Killed`` thinking it was some kind of "game is running" safeguard.

ubitux replied:
I kept the existing one in INotifyKilled.Killed notification (see #6223) and removed in the 3 I added (INotifyCreated.Created, INotifyOwnerChanged.OnOwnerChanged and INotifyActorDisposing.Disposing).

ubitux commented on OpenRA.Mods.Common/Traits/Player/PlayerStatistics.cs:
```diff
@@ -128,12 +161,41 @@ void INotifyKilled.Killed(Actor self, AttackInfo e)
+			if (self.Owner.WinState != WinState.Undefined)
+				return;
+
+			var oldOwnerStats = oldOwner.PlayerActor.Trait<PlayerStatistics>();
```
I'm a bit uncomfortable about leaving the oldOwner parameter unused but ok

ubitux commented on OpenRA.Mods.Common/Traits/Player/PlayerStatistics.cs:
```diff
@@ -128,12 +161,41 @@ void INotifyKilled.Killed(Actor self, AttackInfo e)
+			if (self.Owner.WinState != WinState.Undefined)
+				return;
+
+			if (info.AddToArmyValue) // set armyValue lazily in case Disposing() is called without Created()
```
I prefered that way but ok, changed.

ubitux commented:
> RA aircraft don't appear to be counted towards army value.

ubitux commented:
This is another preview of the army value tab in RA:

![2018-12-27-230307_1041x516_scrot](https://user-images.githubusercontent.com/34467/50495806-d2eb0580-0a2b-11e9-92d2-0a38442f68a7.png)

Earning tab of the same game for comparison:

![2018-12-27-230342_1027x503_scrot](https://user-images.githubusercontent.com/34467/50495834-ff068680-0a2b-11e9-8b20-5ce5ebafea98.png)

The army tab provides information for analysis about important turning points in the game, typically related to bad or good trades.

matjaeck commented:
Would be cool if this made it into the playtest. Some comments:

- MCV: I think it's an army asset in RA due to its basepush potential. To avoid inconsistency I suggest to include it for all default mods and TS.
- Could we have the total army value next to the assets value in the economy tab, too? Then we had all important indicators available in one view.

ubitux commented:
> Would be cool if this made it into the playtest.

I agree ;)

> MCV: I think it's an army asset in RA due to its basepush potential. To avoid inconsistency I suggest to include it for all default mods and TS.

Not so fond of this: are we going to loose $2000 of army value when it transforms? I'll wait for more opinions before changing this.

> Could we have the total army value next to the assets value in the economy tab, too? Then we had all important indicators available in one view.

It's present in the combat tab already; do you think it should be moved? copied?

pchote commented:
> Not so fond of this: are we going to loose $2000 of army value when it transforms?

I agree, better to not include the MCV.

matjaeck commented:
> Not so fond of this: are we going to loose $2000 of army value when it transforms?

Good point.

> It's present in the combat tab already; do you think it should be moved? copied?

Didn't know, will test the PR later. I think there will be different opinions about it so let's just leave it as it is and perhaps adjust it in a follow up PR.

Smittytron reviewed (dismissed):
Can't speak for the code but this looks good in game.

We can gather feedback about whether or not engis, medics and mechanics belong over the playtest and change if needed.

pchote commented on mods/cnc/rules/defaults.yaml:
```diff
@@ -293,6 +295,8 @@
 	Inherits@1: ^ExistsInWorld
 	Inherits@3: ^SpriteActor
+	UpdatesPlayerStatistics:
+		AddToArmyValue: true
```
If we decided that chinooks don't count then they will need to be disabled in this mod too.

pchote commented on mods/cnc/rules/defaults.yaml:
```diff
@@ -651,6 +659,8 @@
 	Inherits@1: ^ExistsInWorld
 	Inherits@3: ^SpriteActor
+	UpdatesPlayerStatistics:
+		AddToArmyValue: true
```
The only ship in TD is the gunboat, which is an unconrtollable support unit like the a10.  Since the a10 doesn't count, this shouldn't either.

ubitux replied:
Dropped.

ubitux replied:
Dropped.

pchote commented on mods/cnc/rules/defaults.yaml:
```diff
@@ -627,6 +633,8 @@
 	Inherits@1: ^ExistsInWorld
 	Inherits@2: ^SpriteActor
+	UpdatesPlayerStatistics:
+		AddToArmyValue: true
```
The only planes in TD are the vehicle production plane and a10 which are both uncontrollable support units.  The c17 definitely shouldn't count, and IMO the a10 shouldn't either.

ubitux replied:
Dropped.

pchote commented on mods/cnc/rules/vehicles.yaml:
```diff
@@ -664,6 +668,8 @@ TRUCK:
 		Cost: 500
 	Tooltip:
 		Name: Supply Truck
+	UpdatesPlayerStatistics:
```
MHQ (line 640 just above) shouldn't count either.

ubitux replied:
Explicitely removed.

pchote commented on mods/d2k/rules/defaults.yaml:
```diff
@@ -337,6 +341,8 @@
 	Inherits@1: ^ExistsInWorld
 	Inherits@2: ^SpriteActor
+	UpdatesPlayerStatistics:
+		AddToArmyValue: true
```
This will include the reinforcement carryall that delivers harvesters when a refinery is placed, the frigate that delivers starport units, and the ornithopter airstrike aircraft - these are all uncontrollable support actors and definitely shouldn't count - and also player-built carryalls, which could be argued either way.

ubitux replied:
Dropped.

pchote commented:
Considering the subtleties revealed in my comments above, I would be happier if we could keep the defaults as they were (`AddToArmyValue` implicitly false) and then explicitly opt in every actor we want to be counted.

pchote commented:
Code changes otherwise look good, so I expect to :+1: and merge as soon as my final points above are resolved one way or another.

ubitux commented on mods/ra/rules/defaults.yaml:
```diff
@@ -542,6 +545,7 @@
 	Armor:
 		Type: Light
 	UpdatesPlayerStatistics:
+		AddToArmyValue: true
```
I admit I just wanted to re-use the `UpdatesPlayerStatistics` field already present here and didn't bother with spy planes and badger bombers because they cost $0 AFAIK. I fixed it in my latest push anyway.

WRT chinooks, I followed the "Army is everything that is not a building, and MCV or a harvester." rule, so they are included, as well as many other units. I think that's fine as is for now.

ubitux commented:
> Considering the subtleties revealed in my comments above, I would be happier if we could keep the defaults as they were (`AddToArmyValue` implicitly false) and then explicitly opt in every actor we want to be counted.

My initial PR was doing that, but the diff was much bigger and hard to review. Following the simpler rule makes IMO more sense here. I can change it back again if this is blocking.

matjaeck commented:
> keep the defaults as they were (AddToArmyValue implicitly false) and then explicitly opt in every actor we want to be counted.

This will make future refinements easier too. I don't think abstraction is very helpful here because feedback/gameplay might require some arbitrary choices and inheritance might have unexpected results.

pchote commented:
> I can change it back again if this is blocking.

I'd prefer it if you could - I can see the inheritance chain causing problems in the future otherwise.
If you'd like to keep a degree of abstraction to reduce boilerplate you could implement the override as its own base template that you can `Inherits@ARMYVALUE: ^AddToArmyValue` at the top of the relevant actors.

ubitux commented:
> > I can change it back again if this is blocking.
> 
> I'd prefer it if you could - I can see the inheritance chain causing problems in the future otherwise.
> If you'd like to keep a degree of abstraction to reduce boilerplate you could implement the override as its own base template that you can `Inherits@ARMYVALUE: ^AddToArmyValue` at the top of the relevant actors.

Nah, it's fine, I'm keeping the 2-lines version since it can be kind of inconsistent with the existing `UpdatesPlayerStatistics` entries.

Branch is updated, but I also kept the old one @ https://github.com/OpenRA/OpenRA/compare/bleed...ubitux:army-value-stats-graph-bak if you change your mind and prefer to merge that one.

```
☭ git diff --stat army-value-stats-graph-bak..army-value-stats-graph
 mods/cnc/rules/aircraft.yaml       |  6 ++++++
 mods/cnc/rules/defaults.yaml       |  6 ------
 mods/cnc/rules/infantry.yaml       | 14 ++++++++++++++
 mods/cnc/rules/ships.yaml          |  2 ++
 mods/cnc/rules/vehicles.yaml       | 32 ++++++++++++++++++++++++--------
 mods/d2k/rules/defaults.yaml       |  4 ----
 mods/d2k/rules/infantry.yaml       | 16 ++++++++++++++++
 mods/d2k/rules/vehicles.yaml       | 24 ++++++++++++++++++++----
 mods/ra/rules/aircraft.yaml        | 10 ++++++++++
 mods/ra/rules/defaults.yaml        |  5 -----
 mods/ra/rules/infantry.yaml        | 32 ++++++++++++++++++++++++++++++++
 mods/ra/rules/ships.yaml           | 12 ++++++++++++
 mods/ra/rules/vehicles.yaml        | 40 ++++++++++++++++++++++++++++++++++------
 mods/ts/rules/aircraft.yaml        | 18 ++++++++++++++++++
 mods/ts/rules/defaults.yaml        |  8 --------
 mods/ts/rules/gdi-infantry.yaml    |  8 ++++++++
 mods/ts/rules/gdi-vehicles.yaml    | 14 ++++++++++++++
 mods/ts/rules/nod-infantry.yaml    |  8 ++++++++
 mods/ts/rules/nod-vehicles.yaml    | 16 ++++++++++++++++
 mods/ts/rules/shared-infantry.yaml |  4 ++++
 mods/ts/rules/shared-vehicles.yaml |  4 ----
 21 files changed, 238 insertions(+), 45 deletions(-)
```

Please double check that I didn't forget something in the process.

pchote commented on OpenRA.Mods.Common/Widgets/Logic/Ingame/ObserverStatsLogic.cs:
```diff
@@ -98,7 +102,8 @@ public class ObserverStatsLogic : ChromeLogic
 				createStatsOption("Combat", combatStatsHeaders, () => DisplayStats(CombatStats)),
-				createStatsOption("Earnings (graph)", earnedThisMinuteGraphHeaders, () => EarnedThisMinuteGraph())
+				createStatsOption("Earnings (graph)", earnedThisMinuteGraphHeaders, () => EarnedThisMinuteGraph()),
+				createStatsOption("Army (graph)", armyThisMinuteGraphHeaders, () => ArmyThisMinuteGraph()),
```
Please also change the `150` in the `statsDropDown.ShowDropDown` line below to `155` to prevent a few px overflow in the dropdown height.

pchote reviewed (commented):
Double checking raised a couple of final things, then lets get this merged.

ubitux commented:
Last changes requested done + bleed rebased
